src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,196 @@
+        /// In case of a doubt, it should respond with true.
+        /// We will be able to cancel later in TryCommit based on VSCompletionItem item, e.g. based on CompletionItemRules.
+        /// </summary>
+        public bool ShouldCommitCompletion(
+            IAsyncCompletionSession session, 
+            SnapshotPoint location, 
+            char typedChar, 
+            CancellationToken cancellationToken)
+            => s_commitChars.Contains(typedChar);
+
+        public AsyncCompletionData.CommitResult TryCommit(
+            IAsyncCompletionSession session, 
+            ITextBuffer subjectBuffer,
+            VSCompletionItem item, 
+            char typeChar, 
+            CancellationToken cancellationToken)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return CommitResultUnhandled;
+            }
+
+            var completionService = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(Source.RoslynItem, out var roslynItem))
+            {
+                // This isn't an item we provided (e.g. Razor). Let the editor handle it normally.
+                return CommitResultUnhandled;
+            }
+
+            // We can be called before for ShouldCommitCompletion. However, that call does not provide rules applied for the completion item.
+            // Now we check for the commit charcter in the context of Rules that could change the list of commit characters.
+            if (!IsCommitCharacter(typeChar, roslynItem.Rules.CommitCharacterRules))
+            {
+                return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.CancelCommit);
+            }
+
+            if (completionService.GetProvider(roslynItem) is IFeaturesCustomCommitCompletionProvider featuresCustomCommitCompletionProvider)
+            {
+                // Custom commit provider assumes that null is provided is case of invoke. VS provides '\0' in the case.
+                char? commitChar = typeChar == '\0' ? null : (char?)typeChar;
+                var commitBehavior = CustomCommit(document, featuresCustomCommitCompletionProvider, session.TextView, 
+                    subjectBuffer, roslynItem, session.ApplicableToSpan, commitChar, cancellationToken);
+
+                return new AsyncCompletionData.CommitResult(isHandled: true, commitBehavior);
+            }
+
+            // TODO Remove language specific code: https://github.com/dotnet/roslyn/issues/30276
+            if (document.Project.Language == LanguageNames.VisualBasic && typeChar == '\n')
+            {
+                return new AsyncCompletionData.CommitResult(isHandled: false, 
+                    AsyncCompletionData.CommitBehavior.RaiseFurtherReturnKeyAndTabKeyCommandHandlers);
```

Comment:
this is incorrect.  this behavior is configurable by the language.  i.e.:

![image](https://user-images.githubusercontent.com/4564579/46449282-0d504980-c740-11e8-9028-572c68314ba4.png)

this will regress VB return behavior for some customers.  We need tests for this.

Reply:
using  Controller.SendEnterThroughToEditor(rules, roslynItem, filterText) now

---
In reply to: [222519999](https://github.com/dotnet/roslyn/pull/29016#discussion_r222519999) [](ancestors = 222519999)